pkg/kv/kvserver/replica_raftlog_test.go

```diff
@@ -0,0 +1,303 @@
+// Copyright 2025 The Cockroach Authors.
+//
+// Use of this software is governed by the CockroachDB Software License
+// included in the /LICENSE file.
+
+package kvserver
+
+import (
+	"context"
+	"math"
+	"math/rand/v2"
+	"testing"
+
+	"github.com/cockroachdb/cockroach/pkg/kv/kvpb"
+	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvserverpb"
+	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/logstore"
+	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/raftentry"
+	"github.com/cockroachdb/cockroach/pkg/raft"
+	"github.com/cockroachdb/cockroach/pkg/raft/raftpb"
+	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
+	"github.com/cockroachdb/cockroach/pkg/storage"
+	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/cockroach/pkg/util/metric"
+	"github.com/cockroachdb/cockroach/pkg/util/stop"
+	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
+	"github.com/stretchr/testify/require"
+)
+
+func TestReplicaLogStorage(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	rt := newReplicaLogStorageTest(t)
+	rt.start(ctx)
+	defer rt.stop(ctx)
+
+	seed := rand.Uint64()
+	t.Logf("seed: %d", seed)
+	rng := newRNG(seed)
+
+	// Start separate threads for truncations and reads under mu. Use stable seed
+	// for each thread.
+	// NB: stable seeds do not achieve reproducibility here, but we try hard. Note
+	// that at least the append thread below is reproducible.
+	for i := 0; i < 5; i++ {
+		fork := newRNG(rng.Uint64())
+		require.NoError(t, rt.stopper.RunAsyncTaskEx(ctx, stop.TaskOpts{TaskName: "read"},
+			func(ctx context.Context) { rt.readLoop(ctx, fork) }))
+	}
+	{
+		fork := newRNG(rng.Uint64())
+		require.NoError(t, rt.stopper.RunAsyncTaskEx(ctx, stop.TaskOpts{TaskName: "trunc"},
+			func(ctx context.Context) { rt.truncateLoop(ctx, fork) }))
+	}
+	// Run the append loop synchronously, and terminate all other tasks when done.
+	rt.writeLoop(ctx, rng, 1000)
+	cancel()
+}
+
+type replicaLogStorageTest struct {
+	t *testing.T
+
+	stopper *stop.Stopper
+
+	ls *replicaLogStorage
+	mu struct {
+		syncutil.RWMutex
+		committed uint64
+		readable  uint64
+	}
+	raftMu syncutil.Mutex
+
+	term uint64
+	ms   *raft.MemoryStorage
+}
+
+func newReplicaLogStorageTest(t *testing.T) *replicaLogStorageTest {
+	const rangeID = 10
+	rt := &replicaLogStorageTest{
+		t:       t,
+		stopper: stop.NewStopper(),
+		term:    1,
+		ms:      raft.NewMemoryStorage(),
+	}
+
+	st := cluster.MakeTestingClusterSettings()
+	eng := storage.NewDefaultInMemForTesting()
+	sideloaded := logstore.NewDiskSideloadStorage(st, rangeID,
+		eng.GetAuxiliaryDir(), nil /* limiter: unused */, eng)
+
+	rt.ls = &replicaLogStorage{
+		ctx:     context.Background(),
+		cache:   raftentry.NewCache(1 << 5), // stress eviction
+		onSync:  noopSyncCallback{},
+		metrics: newStoreMetrics(metric.TestSampleInterval),
```

Comment:
Looks like this is near the only field preventing `replicaLogStorage` to move to a subpackage? Exciting!

Reply:
Yeah, pretty much. There is also a bunch of direct access to `replicaLogStorage` fields, but hopefully there is enough structure there to wrap it into the type's API. Or the fields can be made public worst case.